README.md
# JMeter AMQP Plugin

[![build status](https://img.shields.io/github/workflow/status/aliesbelik/jmeter-amqp-plugin/CI?label=build&color=18BA91&style=flat-square)](https://github.com/aliesbelik/jmeter-amqp-plugin/actions/workflows/maven.yml)
[![release](https://img.shields.io/github/v/release/aliesbelik/jmeter-amqp-plugin?label=release&color=18BA91&style=flat-square)](https://github.com/aliesbelik/jmeter-amqp-plugin/releases/latest)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?color=18BA91&style=flat-square)](https://github.com/aliesbelik/jmeter-amqp-plugin/issues)
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Faliesbelik%2Fjmeter-amqp-plugin.svg?type=shield)](https://app.fossa.com/projects/git%2Bgithub.com%2Faliesbelik%2Fjmeter-amqp-plugin?ref=badge_shield)

A [JMeter](http://jmeter.apache.org/) plugin to publish & consume messages from [RabbitMQ](http://www.rabbitmq.com/)
or any [AMQP](http://www.amqp.org/) message broker.

## Requirements

  - Requires JDK 8 or higher.
  - Compatible with versions up to **5.15.x** of [com.rabbitmq:amqp-client](https://github.com/rabbitmq/rabbitmq-java-client).

## Installation

1. Download the [latest release](https://github.com/aliesbelik/jmeter-amqp-plugin/releases/latest) or
[build your own](#build) from the source code.
2. Put `jmeter-amqp-plugin-VERSION.jar` into `$JMETER_HOME/lib/ext` directory.\
It contains the necessary AMQP client dependency, no other jars are required.
3. Restart JMeter GUI.

## Usage

This plugin includes 2 samplers:

  - [**AMQP Publisher**](#amqp-publisher)
  - [**AMQP Consumer**](#amqp-consumer)

Both samplers can be found in the Add -> Sampler list shown as below:

![amqp-plugin-samplers](docs/images/amqp-plugin-samplers.png)

### AMQP Publisher

![amqp-publisher](docs/images/amqp-publisher.png)

### AMQP Consumer

![amqp-consumer](docs/images/amqp-consumer.png)

A pre-configured example scenario can be found in the [examples](docs/examples) directory.

## Build

### Build requirements

In order to build JMeter AMQP plugin from source, you will need:

  - [Java 8 or higher](https://www.oracle.com/downloads/)
  - [Apache Maven 3](https://maven.apache.org/)

Build dependencies are managed by Maven.\
Jars should automatically be downloaded by Maven as part of the build process.

In addition, you'll need to copy or symlink the following from `$JMETER_HOME/lib/ext` directory:

  - ApacheJMeter_core.jar

### Build from source

To build, just execute:

```
mvn clean package
```

This will create 2 jars in `/target` directory - the jar with all the dependencies within
(this is the one you need) and the original jar (without amqp-client library inside):

```
target/jmeter-amqp-plugin-VERSION.jar
target/original-jmeter-amqp-plugin-VERSION.jar
```

To install, execute:

```
cp target/jmeter-amqp-plugin-VERSION.jar $JMETER_HOME/lib/ext
```

## Acknowledgements

Thanks to

  - [jlavallee](https://github.com/jlavallee) (Jeff Lavallee), who created [original plugin](https://github.com/jlavallee/JMeter-Rabbit-AMQP);
  - [polarnik](https://github.com/polarnik) (Viacheslav Smirnov), who spent time and efforts to create
[comparison](https://github.com/polarnik/JMeter-Rabbit-AMQP-info) of the forks with valuable changes;
  - for the features & enhancements implemented in their forks:
    - [gregLibert](https://github.com/gregLibert/JMeter-Rabbit-AMQP) (Gregory Libert),
    - [gybandi](https://github.com/gybandi/JMeter-Rabbit-AMQP) (Gyuro Andras),
    - [looseend](https://github.com/looseend/JMeter-Rabbit-AMQP) (John Singleton),
    - [sergiogouveia](https://github.com/sergiogouveia/JMeter-Rabbit-AMQP) (Sergio Gouveia),
    - [wojtek-oledzki](https://github.com/voytek-solutions/JMeter-Rabbit-AMQP) (Wojtek Oledzki),
    - [zeph1rus](https://github.com/zeph1rus/JMeter-Rabbit-AMQP);
  - all the [contributors](https://github.com/jlavallee/JMeter-Rabbit-AMQP/graphs/contributors) to original project.

## Contributing

Contributions are welcome.\
[Raise an issue](https://github.com/aliesbelik/jmeter-amqp-plugin/issues), [open a pull request](https://github.com/aliesbelik/jmeter-amqp-plugin/pulls),
or [start a discussion](https://github.com/aliesbelik/jmeter-amqp-plugin/discussions) with your ideas.


## License
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Faliesbelik%2Fjmeter-amqp-plugin.svg?type=large)](https://app.fossa.com/projects/git%2Bgithub.com%2Faliesbelik%2Fjmeter-amqp-plugin?ref=badge_large)

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

* App_id message property for AMQP Publisher (see [jlavallee#37](https://github.com/jlavallee/JMeter-Rabbit-AMQP/issues/37)).
* Timestamp message property for AMQP Publisher (see [jlavallee#37](https://github.com/jlavallee/JMeter-Rabbit-AMQP/issues/37)).

### Dependency Updates

* Up amqp-client version to 5.16.0.
* Up slf4j-api to 2.0.1.

## [0.2.0] - 2022-08-02

Update RabbitMQ amqp-client from 4.x to 5.x.
Look into [changes between 4.x.x and 5.0.0](https://github.com/rabbitmq/rabbitmq-java-client/releases/tag/v5.0.0) for details.

### Added

* CodeQL and SonarCloud analysis.

### Changed

* RabbitMQ amqp-client compatibility bumped from 4.x.x to 5.x.x [#8](https://github.com/aliesbelik/jmeter-amqp-plugin/pull/8). Thanks to [@t-h-e](https://github.com/t-h-e).
* Single jar with dependencies provided as release artifact [#1](https://github.com/aliesbelik/jmeter-amqp-plugin/pull/1).

### Fixed

* NullPointerException on purging queue in AMQP Consumer [#10](https://github.com/aliesbelik/jmeter-amqp-plugin/pull/10). Thanks to [@t-h-e](https://github.com/t-h-e).

### Dependency Updates

* Up jmeter-core version to 5.5.
* Up amqp-client version to 5.15.0.

## [0.1.0] - 2022-01-15

Initial release.

### Added

* Content-encoding message property (for AMQP Publisher). Thanks to [@gybandi](https://github.com/gybandi).
* Queue x-max-priority & message priority parameters (see [jlavallee#42](https://github.com/jlavallee/JMeter-Rabbit-AMQP/pull/42)). Thanks to [@gregLibert](https://github.com/gregLibert), [@looseend](https://github.com/looseend).
* Request (for AMQP Publisher) and response (for AMQP Consumer) headers.
* Option to configure transactional AMQP consumer from the UI (see [jlavallee#38](https://github.com/jlavallee/JMeter-Rabbit-AMQP/pull/38)). Thanks to [@nicklasbondesson](https://github.com/nicklasbondesson).
* Auto-delete property for exchange (see [jlavallee#33](https://github.com/jlavallee/JMeter-Rabbit-AMQP/pull/33)). Thanks to [@wneild](https://github.com/wneild).
* Heartbeat option (see [zeph1rus/JMeter-Rabbit-AMQP@efddefa](https://github.com/zeph1rus/JMeter-Rabbit-AMQP/commit/efddefad62aa54eed4a96dd4cc0b9fe2fb040e1a)). Thanks to [@zeph1rus](https://github.com/zeph1rus).
* Jar with dependencies.
* Extra exception handling.

### Changed

* Maven instead of ant & ivy.
* Switch to maven directory layout.
* Switch to SLF4J API for logging.
* Samplers configuration screens adjusted.
* Scenario example updated.
* Code cleanup.

### Fixed

* No header in reply-to message issue (see [jlavallee#41](https://github.com/jlavallee/JMeter-Rabbit-AMQP/issues/41)). Thanks to [@gregLibert](https://github.com/gregLibert).
* Checkstyle issues.

### Removed

* Redundant dependencies removed.

### Dependency Updates

* Up jmeter-core version to 5.4.3.
* Up amqp-client version to 4.12.0.
* Up commons-lang3 version to 3.12.0.

[Unreleased]: https://github.com/aliesbelik/jmeter-amqp-plugin/compare/v0.2.0...HEAD
[0.2.0]: https://github.com/aliesbelik/jmeter-amqp-plugin/releases/tag/v0.2.0
[0.1.0]: https://github.com/aliesbelik/jmeter-amqp-plugin/releases/tag/v0.1.0


src/main/java/com/zeroclue/jmeter/protocol/amqp/AMQPConsumer.java
package com.zeroclue.jmeter.protocol.amqp;

import com.rabbitmq.client.Channel;
import com.rabbitmq.client.ConsumerCancelledException;
import com.rabbitmq.client.DeliverCallback;
import com.rabbitmq.client.Delivery;
import com.rabbitmq.client.ShutdownSignalException;

import org.apache.groovy.util.Maps;
import org.apache.jmeter.samplers.Entry;
import org.apache.jmeter.samplers.Interruptible;
import org.apache.jmeter.samplers.SampleResult;
import org.apache.jmeter.testelement.TestStateListener;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.security.KeyManagementException;
import java.security.NoSuchAlgorithmException;
import java.util.Map;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class AMQPConsumer extends AMQPSampler implements Interruptible, TestStateListener {

    private static final long serialVersionUID = 7480863561320459091L;

    private static final Logger log = LoggerFactory.getLogger(AMQPConsumer.class);

    private static final Map<Class<? extends Exception>, String> EXCEPTION_TO_RESPONSE_CODE = Maps.of(
        IOException.class, "100",
        InterruptedException.class, "200",
        ConsumerCancelledException.class, "300",
        ShutdownSignalException.class, "400"
    );

    //++ These are JMX names, and must not be changed
    private static final String PREFETCH_COUNT          = "AMQPConsumer.PrefetchCount";
    private static final String READ_RESPONSE           = "AMQPConsumer.ReadResponse";
    private static final String PURGE_QUEUE             = "AMQPConsumer.PurgeQueue";
    private static final String AUTO_ACK                = "AMQPConsumer.AutoAck";
    private static final String RECEIVE_TIMEOUT         = "AMQPConsumer.ReceiveTimeout";
    private static final String USE_TX                  = "AMQPConsumer.UseTx";

    public static final String TIMESTAMP_PARAMETER      = "Timestamp";
    public static final String EXCHANGE_PARAMETER       = "Exchange";
    public static final String ROUTING_KEY_PARAMETER    = "Routing Key";
    public static final String DELIVERY_TAG_PARAMETER   = "Delivery Tag";
    public static final String APP_ID_PARAMETER         = "Application ID";

    public static final boolean DEFAULT_PURGE_QUEUE = false;
    public static final boolean DEFAULT_AUTO_ACK = true;
    public static final boolean DEFAULT_READ_RESPONSE = true;
    public static final boolean DEFAULT_USE_TX = false;
    private static final int DEFAULT_PREFETCH_COUNT = 0;    // unlimited
    public static final String DEFAULT_PREFETCH_COUNT_STRING = Integer.toString(DEFAULT_PREFETCH_COUNT);
    public static final String DEFAULT_RESPONSE_CODE = "500";
    public static final String DEFAULT_RECEIVE_TIMEOUT = "";

    private transient Channel channel;
    private transient DeliverCallback consumer;
    private transient BlockingQueue<Delivery> response;
    private transient String consumerTag;

    public AMQPConsumer() {
        super();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public SampleResult sample(Entry entry) {
        SampleResult result = new SampleResult();
        result.setSampleLabel(getName());
        result.setSuccessful(false);
        result.setResponseCode(DEFAULT_RESPONSE_CODE);
        result.setSampleLabel(getTitle());

        trace("AMQPConsumer.sample()");

        try {
            initChannel();

            if (purgeQueue()) {
                doPurgeQueue();
            }

            // only do this once per thread, otherwise it slows down the consumption by appx 50%
            if (consumer == null) {
                log.info("Creating consumer");
                response = new LinkedBlockingQueue<>(1);
                consumer = (consumerTag, delivery) -> response.offer(delivery);
            }
            if (consumerTag == null) {
                log.info("Starting basic consumer");
                consumerTag = channel.basicConsume(getQueue(), autoAck(), consumer, consumerTag  -> { });
            }
        } catch (Exception ex) {
            log.error("Failed to initialize channel", ex);
            result.setResponseMessage(ex.toString());
            return result;
        }

        /*
         * Perform the sampling
         */

        // aggregate samples
        int loop = getIterationsAsInt();
        result.sampleStart();                      // start timing
        Delivery delivery = null;

        try {
            for (int idx = 0; idx < loop; idx++) {
                delivery = response.poll(getReceiveTimeoutAsInt(), TimeUnit.MILLISECONDS);

                if (delivery == null) {
                    result.setResponseMessage("Timed out");
                    return result;
                }

                /*
                 * Set up the sample result details
                 */
                if (getReadResponseAsBoolean()) {
                    String responseStr = new String(delivery.getBody());
                    result.setSamplerData(responseStr);
                    result.setResponseData(responseStr, null);
                } else {
                    result.setSamplerData("Read response is false.");
                }

                if (!autoAck()) {
                    channel.basicAck(delivery.getEnvelope().getDeliveryTag(), false);
                }
            }

            // commit the sample
            if (getUseTx()) {
                channel.txCommit();
            }

            /*
             * Set up the sample result details
             */
            result.setResponseMessage("OK");
            result.setDataType(SampleResult.TEXT);
            result.setResponseHeaders(delivery != null ? formatHeaders(delivery) : null);

            result.setResponseCodeOK();
            result.setSuccessful(true);
        } catch(InterruptedException ie) {
            Thread.currentThread().interrupt();     // re-interrupt the current thread
            response = null;
            consumer = null;
            consumerTag = null;
            log.warn("Interrupted while attempting to consume", ie);
            result.setResponseCode(EXCEPTION_TO_RESPONSE_CODE.get(ie.getClass()));
            result.setResponseMessage(ie.getMessage());
        } catch (ShutdownSignalException | ConsumerCancelledException | IOException e) {
            response = null;
            consumer = null;
            consumerTag = null;
            log.warn("AMQP consumer failed to consume", e);
            result.setResponseCode(EXCEPTION_TO_RESPONSE_CODE.get(e.getClass()));
            result.setResponseMessage(e.getMessage());
            interrupt();
        } finally {
            result.sampleEnd();         // end timing
        }

        trace("AMQPConsumer.sample ended");

        return result;
    }

    @Override
    protected Channel getChannel() {
        return channel;
    }

    @Override
    protected void setChannel(Channel channel) {
        this.channel = channel;
    }

    /**
     * @return the whether to purge the queue
     */
    public String getPurgeQueue() {
        return getPropertyAsString(PURGE_QUEUE);
    }

    public void setPurgeQueue(String content) {
        setProperty(PURGE_QUEUE, content);
    }

    public void setPurgeQueue(Boolean purgeQueue) {
        setProperty(PURGE_QUEUE, purgeQueue.toString());
    }

    public boolean purgeQueue() {
        return Boolean.parseBoolean(getPurgeQueue());
    }

    private void doPurgeQueue() {
        log.info("Purging queue {}", getQueue());
        try {
            channel.queuePurge(getQueue());
        } catch (IOException e) {
            log.error("Failed to purge queue {}", getQueue(), e);
        }
    }

    /**
     * @return the whether to auto ack
     */
    public String getAutoAck() {
        return getPropertyAsString(AUTO_ACK);
    }

    public void setAutoAck(String content) {
        setProperty(AUTO_ACK, content);
    }

    public void setAutoAck(Boolean autoAck) {
        setProperty(AUTO_ACK, autoAck.toString());
    }

    public boolean autoAck() {
        return getPropertyAsBoolean(AUTO_ACK);
    }

    protected int getReceiveTimeoutAsInt() {
        if (getPropertyAsInt(RECEIVE_TIMEOUT) < 1) {
            return DEFAULT_TIMEOUT;
        }

        return getPropertyAsInt(RECEIVE_TIMEOUT);
    }

    public String getReceiveTimeout() {
        return getPropertyAsString(RECEIVE_TIMEOUT, DEFAULT_TIMEOUT_STRING);
    }

    public void setReceiveTimeout(String s) {
        setProperty(RECEIVE_TIMEOUT, s);
    }

    public String getPrefetchCount() {
        return getPropertyAsString(PREFETCH_COUNT, DEFAULT_PREFETCH_COUNT_STRING);
    }

    public void setPrefetchCount(String prefetchCount) {
        setProperty(PREFETCH_COUNT, prefetchCount);
    }

    public int getPrefetchCountAsInt() {
        return getPropertyAsInt(PREFETCH_COUNT);
    }

    public boolean getUseTx() {
        return getPropertyAsBoolean(USE_TX, DEFAULT_USE_TX);
    }

    public void setUseTx(Boolean tx) {
        setProperty(USE_TX, tx);
    }

    /**
     * Option if the sampler should read the response.
     *
     * @return whether the sampler should read the response
     */
    public String getReadResponse() {
        return getPropertyAsString(READ_RESPONSE);
    }

    /**
     * Set option if the sampler should read the response.
     *
     * @param read whether the sampler should read the response or not
     */
    public void setReadResponse(Boolean read) {
        setProperty(READ_RESPONSE, read);
    }

    /**
     * Option if the sampler should read the response as a boolean value.
     *
     * @return whether the sampler should read the response as a boolean value
     */
    public boolean getReadResponseAsBoolean() {
        return getPropertyAsBoolean(READ_RESPONSE);
    }

    @Override
    public boolean interrupt() {
        testEnded();
        return true;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void testEnded() {
        // Do nothing
    }

    @Override
    public void testEnded(String arg0) {
        // Do nothing
    }

    @Override
    public void testStarted() {
        // Do nothing
    }

    @Override
    public void testStarted(String arg0) {
        // Do nothing
    }

    @Override
    public void cleanup() {
        try {
            if (consumerTag != null) {
                channel.basicCancel(consumerTag);
            }
        } catch (IOException e) {
            log.error("Couldn't safely cancel the sample {}", consumerTag, e);
        }

        super.cleanup();
    }

    /**
     * Helper method.
     */
    private void trace(String s) {
        String tl = getTitle();
        String tn = Thread.currentThread().getName();
        String th = this.toString();
        log.debug("{} {} {} {}", tn, tl, s, th);
    }

    @Override
    protected boolean initChannel() throws IOException, NoSuchAlgorithmException, KeyManagementException, TimeoutException {
        boolean ret = super.initChannel();
        channel.basicQos(getPrefetchCountAsInt());

        if (getUseTx()) {
            channel.txSelect();
        }

        return ret;
    }

    private String formatHeaders(Delivery delivery) {
        Map<String, Object> headers = delivery.getProperties().getHeaders();
        StringBuilder sb = new StringBuilder();

        if (delivery.getProperties().getTimestamp() != null) {
            sb.append(TIMESTAMP_PARAMETER)
                .append(": ")
                .append((delivery.getProperties().getTimestamp().getTime())/1000)
                .append("\n");
        }

        sb.append(EXCHANGE_PARAMETER)
            .append(": ")
            .append(delivery.getEnvelope().getExchange())
            .append("\n");
        sb.append(ROUTING_KEY_PARAMETER)
            .append(": ")
            .append(delivery.getEnvelope().getRoutingKey())
            .append("\n");
        sb.append(DELIVERY_TAG_PARAMETER)
            .append(": ")
            .append(delivery.getEnvelope().getDeliveryTag())
            .append("\n");

        if (delivery.getProperties().getAppId() != null) {
            sb.append(APP_ID_PARAMETER)
                .append(": ")
                .append(delivery.getProperties().getAppId())
                .append("\n");
        }

        if (headers != null) {
            for (Map.Entry<String,Object> entry : headers.entrySet()) {
                sb.append(entry.getKey())
                    .append(": ")
                    .append(entry.getValue())
                    .append("\n");
            }
        }

        return sb.toString();
    }
}


src/main/java/com/zeroclue/jmeter/protocol/amqp/gui/AMQPHeadersPanel.java
package com.zeroclue.jmeter.protocol.amqp.gui;

import org.apache.jmeter.config.Argument;
import org.apache.jmeter.config.gui.ArgumentsPanel;
import org.apache.jorphan.gui.ObjectTableModel;
import org.apache.jorphan.reflect.Functor;

public class AMQPHeadersPanel extends ArgumentsPanel {

    public AMQPHeadersPanel(final String label) {
        super(label);
    }

    @Override
    protected void initializeTableModel() {
        tableModel = new ObjectTableModel(new String[] { COLUMN_RESOURCE_NAMES_0, COLUMN_RESOURCE_NAMES_1, "Type" },
            Argument.class,
            new Functor[] {
                new Functor("getName"), // $NON-NLS-1$
                new Functor("getValue"),  // $NON-NLS-1$
                new Functor("getDescription") },  // $NON-NLS-1$
            new Functor[] {
                new Functor("setName"), // $NON-NLS-1$
                new Functor("setValue"), // $NON-NLS-1$
                new Functor("setDescription") },  // $NON-NLS-1$
            new Class[] { String.class, String.class, String.class });
    }

    @Override
    protected Argument makeNewArgument() {
        return new Argument("", "", null, "string");
    }

}


src/main/java/com/zeroclue/jmeter/protocol/amqp/gui/AMQPPublisherGui.java
package com.zeroclue.jmeter.protocol.amqp.gui;

import com.zeroclue.jmeter.protocol.amqp.AMQPPublisher;

import java.awt.Dimension;
import java.awt.GridBagConstraints;
import java.awt.GridBagLayout;

import javax.swing.BorderFactory;
import javax.swing.JCheckBox;
import javax.swing.JPanel;
import javax.swing.SwingConstants;

import org.apache.jmeter.config.Arguments;
import org.apache.jmeter.config.gui.ArgumentsPanel;
import org.apache.jmeter.gui.util.HorizontalPanel;
import org.apache.jmeter.testelement.TestElement;
import org.apache.jorphan.gui.JLabeledTextArea;
import org.apache.jorphan.gui.JLabeledTextField;

/**
 * GUI for AMQP Publisher.
 */
public class AMQPPublisherGui extends AMQPSamplerGui {

    private static final long serialVersionUID = 1L;

    private final JLabeledTextArea message = new JLabeledTextArea("Message Content");

    private final JLabeledTextField messageRoutingKey = new JLabeledTextField("          Routing Key");
    private final JLabeledTextField messageType = new JLabeledTextField("     Message Type");
    private final JLabeledTextField replyToQueue = new JLabeledTextField("   Reply-To Queue");
    private final JLabeledTextField messageId = new JLabeledTextField("           Message ID");
    private final JLabeledTextField correlationId = new JLabeledTextField("       Correlation ID");
    private final JLabeledTextField messagePriority = new JLabeledTextField(" Message Priority");
    private final JLabeledTextField contentType = new JLabeledTextField("        Content-Type");
    private final JLabeledTextField contentEncoding = new JLabeledTextField("Content Encoding");
    private final JLabeledTextField appId = new JLabeledTextField("       Application ID");

    private final JCheckBox timestamp = new JCheckBox("Timestamp", AMQPPublisher.DEFAULT_TIMESTAMP);
    private final JCheckBox persistent = new JCheckBox("Persistent", AMQPPublisher.DEFAULT_PERSISTENT);
    private final JCheckBox useTx = new JCheckBox("Use Transactions", AMQPPublisher.DEFAULT_USE_TX);

    private final ArgumentsPanel headers = new ArgumentsPanel("Headers");

    private static final String MSG_SETTINGS_LABEL = "Message";
    private static final String PROPS_SETTINGS_LABEL = "Properties";

    private JPanel mainPanel;

    public AMQPPublisherGui() {
        init();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String getLabelResource() {
        return this.getClass().getSimpleName();
    }

    @Override
    public String getStaticLabel() {
        return "AMQP Publisher";
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void configure(TestElement element) {
        super.configure(element);
        if (!(element instanceof AMQPPublisher)) {
            return;
        }
        AMQPPublisher sampler = (AMQPPublisher) element;

        persistent.setSelected(sampler.getPersistent());
        useTx.setSelected(sampler.getUseTx());

        messageRoutingKey.setText(sampler.getMessageRoutingKey());
        messageType.setText(sampler.getMessageType());
        replyToQueue.setText(sampler.getReplyToQueue());
        contentType.setText(sampler.getContentType());
        contentEncoding.setText(sampler.getContentEncoding());
        correlationId.setText(sampler.getCorrelationId());
        messagePriority.setText(sampler.getMessagePriority());
        messageId.setText(sampler.getMessageId());
        message.setText(sampler.getMessage());
        appId.setText(sampler.getAppId());
        timestamp.setSelected(sampler.getTimestamp());

        configureHeaders(sampler);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public TestElement createTestElement() {
        AMQPPublisher sampler = new AMQPPublisher();
        modifyTestElement(sampler);

        return sampler;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void modifyTestElement(TestElement te) {
        AMQPPublisher sampler = (AMQPPublisher) te;
        sampler.clear();
        configureTestElement(sampler);

        super.modifyTestElement(sampler);

        sampler.setPersistent(persistent.isSelected());
        sampler.setUseTx(useTx.isSelected());

        sampler.setMessageRoutingKey(messageRoutingKey.getText());
        sampler.setMessage(message.getText());
        sampler.setMessageType(messageType.getText());
        sampler.setReplyToQueue(replyToQueue.getText());
        sampler.setCorrelationId(correlationId.getText());
        sampler.setMessagePriority(messagePriority.getText());
        sampler.setContentType(contentType.getText());
        sampler.setContentEncoding(contentEncoding.getText());
        sampler.setMessageId(messageId.getText());
        sampler.setAppId(appId.getText());
        sampler.setTimestamp(timestamp.isSelected());

        sampler.setHeaders((Arguments) headers.createTestElement());
    }

    @Override
    protected void setMainPanel(JPanel panel) {
        mainPanel = panel;
    }

    /*
     * Helper method to set up the GUI screen
     */
    @Override
    protected final void init() {
        super.init();

        mainPanel.add(initMessagePanel());
    }

    private JPanel initMessagePanel() {
        GridBagConstraints constraints = new GridBagConstraints();
        constraints.insets = new java.awt.Insets(2, 2, 2, 2);
        constraints.fill = GridBagConstraints.HORIZONTAL;
        constraints.anchor = GridBagConstraints.LINE_START;
        constraints.weightx = 1.0;
        constraints.weighty = 0.5;
        constraints.gridwidth = GridBagConstraints.REMAINDER;

        JPanel messagePanel = new JPanel(new GridBagLayout());
        messagePanel.setBorder(BorderFactory.createTitledBorder(BorderFactory.createEtchedBorder(), MSG_SETTINGS_LABEL));

        JPanel optionsPanel = new HorizontalPanel();
        optionsPanel.add(persistent);
        optionsPanel.add(useTx);
        messagePanel.add(optionsPanel, constraints);

        messagePanel.add(initMessagePropertyPanel(), constraints);
        message.setPreferredSize(new Dimension(400, 200));
        messagePanel.add(message, constraints);
        messagePanel.add(headers, constraints);

        return messagePanel;
    }

    private JPanel initMessagePropertyPanel() {
        GridBagConstraints constraints = new GridBagConstraints();
        constraints.insets = new java.awt.Insets(2, 2, 2, 2);
        constraints.fill = GridBagConstraints.HORIZONTAL;
        constraints.anchor = GridBagConstraints.LINE_START;
        constraints.weightx = 1.0;
        constraints.weighty = 0.5;
        constraints.gridwidth = GridBagConstraints.REMAINDER;

        JPanel propertyPanel = new JPanel(new GridBagLayout());
        propertyPanel.setBorder(BorderFactory.createTitledBorder(BorderFactory.createEtchedBorder(), PROPS_SETTINGS_LABEL));

        timestamp.setHorizontalTextPosition(SwingConstants.LEFT);
        //timestamp.setHorizontalTextPosition(SwingConstants.LEADING);
        timestamp.setIconTextGap(35);

        propertyPanel.add(messageRoutingKey, constraints);
        propertyPanel.add(replyToQueue, constraints);
        propertyPanel.add(messageType, constraints);
        propertyPanel.add(correlationId, constraints);
        propertyPanel.add(messageId, constraints);
        propertyPanel.add(messagePriority, constraints);
        propertyPanel.add(appId, constraints);
        propertyPanel.add(contentType, constraints);
        propertyPanel.add(contentEncoding, constraints);
        propertyPanel.add(timestamp, constraints);

        return propertyPanel;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void clearGui() {
        super.clearGui();

        persistent.setSelected(AMQPPublisher.DEFAULT_PERSISTENT);
        useTx.setSelected(AMQPPublisher.DEFAULT_USE_TX);
        messageRoutingKey.setText("");
        messageType.setText("");
        replyToQueue.setText("");
        correlationId.setText("");
        messagePriority.setText("");
        contentType.setText(AMQPPublisher.DEFAULT_CONTENT_TYPE);
        contentEncoding.setText(AMQPPublisher.DEFAULT_ENCODING);
        messageId.setText("");
        message.setText("");
        appId.setText("");
        timestamp.setSelected(AMQPPublisher.DEFAULT_TIMESTAMP);

        headers.clearGui();
    }

    private void configureHeaders(AMQPPublisher sampler) {
        Arguments sampleHeaders = sampler.getHeaders();
        if (sampleHeaders != null) {
            headers.configure(sampleHeaders);
        } else {
            headers.clearGui();
        }
    }
}
